### PR TITLE
HashTable Implementation

### PR DESCRIPTION
**Pull request by spierce7**
This is the implementation that @dalewking recommended here: https://github.com/tbroyer/bullet/issues/2

This implementation scales better with large number of injects. This also resolves some bugs I was encountering in the current implementation where it was generating the `instanceof`s out of order and it was calling the parent class inject method instead of the proper child class.

### CODE REVIEW

**Comment by tbroyer** [comment]
> This also resolves some bugs I was encountering in the current implementation where it was generating the instanceofs out of order and it was calling the parent class inject method instead of the proper child class.

Are you sure the fix isn't just the difference between `master` and the last released code? (specifically 3726e0c81c49fa6f1da4edb1c26740138ae3716c)

**Comment by spierce7** [comment]
Ah, I was wondering why the sorting code I saw wasn't working correctly. That would probably fix the issue also. I was using the latest release while encountering the bug.

File: compiler/src/main/java/bullet/impl/ComponentMethodDescriptor.java [hunk 1]
```diff
@@ -1,5 +1,11 @@
 package bullet.impl;
 
+import com.google.auto.common.AnnotationMirrors;
```
Comment (tbroyer) [inline compiler/src/main/java/bullet/impl/ComponentMethodDescriptor.java]: Please don't reorder imports.

http://www.contribution-guide.org/#code-formatting

Comment (spierce7) [inline compiler/src/main/java/bullet/impl/ComponentMethodDescriptor.java]: IntelliJ :-( I'll fix it.

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 2]
```diff
@@ -145,6 +150,15 @@ private String getName(TypeMirror type) {
 
     final ClassName elementName = ClassName.get(element);
 
+    // Filter out duplicate injections
```
Comment (tbroyer) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: If you have several methods injecting the same type, then you're doing it wrong, and it shouldn't matter if the code is suboptimal (or in this case which exact method, and which kind of method is used).

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 3]
```diff
@@ -152,15 +166,33 @@ private String getName(TypeMirror type) {
...
             .addParameter(elementName, "component", FINAL)
             .addCode("this.component = component;\n")
             .build());
 
+    if (membersInjectionMethodsMap.size() > 0) {
+      classBuilder.addField(ClassName.get("bullet.impl", "ClassIndexHashTable"), "classIndexHashTable", PRIVATE, STATIC, FINAL);
+    }
+
+    // Generate the ClassIndexHashTable if there are classes to inject.
+    if (membersInjectionMethodsMap.size() > 0) {
```
Comment (tbroyer) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: That second `if` is not needed: you could just merge the two `if`s together as they have the same condition.

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 4]
```diff
@@ -175,8 +207,7 @@ private String getName(TypeMirror type) {
           "$<}\n",
           method.type(), method.name(), method.kind() == ComponentMethodKind.PROVIDER_OR_LAZY ? ".get()" : "");
     }
-    // TODO: exception message
-    getBuilder.addCode("throw new $T();\n", IllegalArgumentException.class);
+    getBuilder.addCode("throw new $T(\"No get or Provides method found for \" + type.getName() + \" in " + elementName.simpleName() + "\");\n", IllegalArgumentException.class);
```
Comment (tbroyer) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: Again an unrelated change.
I also disagree with the wording; "Provides methods" live in modules; I suppose you meant "methods that return a `Provider`".

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 5]
```diff
@@ -185,16 +216,34 @@ private String getName(TypeMirror type) {
...
+    // Generate injection code if there are injections
+    if (membersInjectionMethodsMap.size() > 0) {
       injectWriter.addCode(
-          "if (instance instanceof $T) {\n$>" +
-          "this.component.$N$L(($T) instance);\n" +
-          "return instance;\n" +
-          "$<}\n",
-          method.type(), method.name(), method.kind() == ComponentMethodKind.MEMBERS_INJECTOR ? "().injectMembers" : "", method.type());
+          "Class<?> c = instance.getClass();\n" +
+              "while (c != Object.class) {\n$>" +
```
Comment (tbroyer) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: Indentation.

http://www.contribution-guide.org/#code-formatting

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 6]
```diff
@@ -185,16 +216,34 @@ private String getName(TypeMirror type) {
...
+              "switch (classIndexHashTable.get(c)) {\n$>");
+
+      {
+        int i = 0;
+        for (Map.Entry<TypeMirror, ComponentMethodDescriptor> entry : membersInjectionMethodsMap.entrySet()) {
+          ComponentMethodDescriptor method = entry.getValue();
+          injectWriter.addCode(
+              "case " + i++ + ":\n$>" +
```
Comment (tbroyer) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: There's `$L` for that.

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 7]
```diff
@@ -185,16 +216,34 @@ private String getName(TypeMirror type) {
...
+
+      injectWriter.addCode(
+          "$<}\n" +
+              "c = c.getSuperclass();\n" +
+              "$<}\n");
     }
-    // TODO: exception message
-    injectWriter.addCode("throw new $T();\n", IllegalArgumentException.class);
+
+    injectWriter.addCode("throw new $T(\"No inject or MembersInject method found for \" + instance.getClass().getName() + \" in " + elementName.simpleName() + "\");\n", IllegalArgumentException.class);
```
Comment (tbroyer) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: `$T`

File: compiler/src/test/java/bullet/impl/AbstractComponentProcessorTest.java [hunk 8]
```diff
@@ -67,35 +68,35 @@
...
-        "    }",
-        "    throw new IllegalArgumentException()",
-        "  }",
-        "",
-        "  @Override",
-        "  public <T> T inject(final T instance) {",
-        "    throw new IllegalArgumentException();",
-        "  }",
-        "}");
+        "package test;\n" +
```
Comment (tbroyer) [inline compiler/src/test/java/bullet/impl/AbstractComponentProcessorTest.java]: Why change the vararg with `\n`‽

File: pom.xml [hunk 9]
```diff
@@ -116,7 +116,7 @@
       <dependency>
         <groupId>com.squareup</groupId>
         <artifactId>javapoet</artifactId>
-        <version>1.0.0</version>
+        <version>1.3.0</version>
```
Comment (tbroyer) [inline pom.xml]: Unrelated change (I suppose, unless it's required by the few code changes above)

Comment (spierce7) [inline pom.xml]: it is.

File: compiler/src/test/java/bullet/impl/AbstractComponentProcessorTest.java (hunk 8 above)
Comment (spierce7) [inline compiler/src/test/java/bullet/impl/AbstractComponentProcessorTest.java]: I'm not sure how you got each of the lines to separate neatly light that. Did you copy each line by hand? IntelliJ has the ability to take code and turn it into a `String`, and this is what it produced.

**Comment by tbroyer** [comment]
I'm not fully opposed to using a map instead of an if-cascade, but the custom map is premature optimization IMO.
I mean, you opted to use Dagger 2 which uses a different paradigm than Dagger 1 (https://speakerdeck.com/jakewharton/dependency-injection-with-dagger-2-devoxx-2014) so you should adapt your code; and it happens that it's not that hard to do. You'd better spend your time eliminating Bullet• from your codebase than optimizing it.

File: compiler/src/test/java/bullet/impl/AbstractComponentProcessorTest.java (hunk 8 above)
Comment (tbroyer) [inline compiler/src/test/java/bullet/impl/AbstractComponentProcessorTest.java]: I hand-wrote the code here. If you just copy/paste the actual produced code, you're not actually testing whether it complies with what you expect.

Comment (spierce7) [inline compiler/src/test/java/bullet/impl/AbstractComponentProcessorTest.java]: I wrote a test class and had IntelliJ turn it into a `String`.

**Comment by spierce7** [comment]
I've seen the talk you linked, as well as the one from Gregory Kick.

I opted to use Dagger 2 as Dagger 1 is no longer supported. Dagger 2 also generates better code, better errors, and integrates well with Proguard. 

That being said: I'm not sure I like the way that Dagger 2 forces it's way into every class that needs injection (which is most of them). I'm unconvinced that it's a better approach. As such, I will continue to hide my injection logic behind a common interface.

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java (hunk 2 above)
Comment (dalewking) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: The issue it that you may have both:

    void inject(Foo f);
    MembersInjector<Foo> getFooInjector();

which is not "doing it wrong" if you have both, you pick the one that is more optimal, which is the members injector.

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 10]
```diff
@@ -152,15 +166,32 @@ private String getName(TypeMirror type) {
...
             .build());
 
+    if (membersInjectionMethodsMap.size() > 0) {
+      classBuilder.addField(ClassName.get("bullet.impl", "ClassIndexHashTable"), "classIndexHashTable", PRIVATE, STATIC, FINAL);
+    }
+
+    // Generate the ClassIndexHashTable if there are classes to inject.
+    if (membersInjectionMethodsMap.size() > 0) {
+      // ClassIndexHashTable size should be at least 30% larger and a prime number.
+      int classIndexHashTableSize = getNextPrime((int) Math.ceil(membersInjectionMethodsMap.size() * 1.3));
```
Comment (dalewking) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: To be precise (and I realize it was me that wasn't precise) you want to achieve a maximum of 70% load factor. Multiplying by 1.3 can be 77% load factor. What I do in my own implementation is divide by 0.7 to get at least a 70% load factor (or you could multiply by 1.43).

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java [hunk 11]
```diff
@@ -224,4 +272,22 @@ private boolean isVisibleFrom(Element target, PackageElement from) {
         throw new AssertionError();
     }
   }
+
+  private static int getNextPrime(int value) {
```
Comment (dalewking) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: BigInteger.nextProbablePrime will do this for you and should be reliable for the small integers we are talking about here.

Comment (spierce7) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: I saw that you were using that. The documentation makes me wary though. I'd rather use something 100% reliable.

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java (hunk 10 above)
Comment (spierce7) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: I appreciate the clarification. Fix commit: https://github.com/WeMakeBetterApps/bullet/commit/05b94baf97df6fea4b5665bba3fb800d0b2328a1

**Comment by bryant1410** [comment]
Where does it say Dagger 1 is no longer supported?

File: compiler/src/main/java/bullet/impl/ComponentProcessingStep.java (hunk 11 above)
Comment (dalewking) [inline compiler/src/main/java/bullet/impl/ComponentProcessingStep.java]: The comments about it not being 100% reliable are due to the fact that it has to handle numbers that can be 1000s of bits long and for really large numbers it isn't 100% reliable. For the range of numbers used here (typically < 100) it will be 100% reliable.
